File: externals/Async-CTP/samples/CS/Testing/Unit/Testing/Test/Library/Mocks/Mock.cs
using System;
using System.Linq.Expressions;
using System.Threading.Tasks;

// This API is designed to mimic Moq, a popular mocking library for .NET
//
// However for this API, we only support mocking IStockSnapshotProvider
//
// This is not how Moq works at all - just the minimal amount of code that allows
// the surface area to appear like Moq.

class Mock<TInterface>
{
    readonly MockSetup mockImpl = new MockSetup();

    internal Mock()
    {
        if (!typeof(TInterface).Equals(typeof(IStockSnapshotProvider)))
        {
            throw new NotImplementedException("This sample only supports mocking IStockSnapshotProvider");
        }
    }

    // Our setup method only supports mocking IStockSnapshotProvider.GetLatestSnapshotAsync()
    internal MockSetup Setup(Expression<Func<IStockSnapshotProvider, Task<StockSnapshot>>> function)
    {
        if (function.Body.NodeType == ExpressionType.Call)
        {
            var innerCall = (MethodCallExpression)function.Body;
            var methodInfo = innerCall.Method;
            if (typeof(IStockSnapshotProvider).Equals(methodInfo.DeclaringType) &&
                methodInfo.IsStatic == false &&
                methodInfo.Name.Equals("GetLatestSnapshotAsync") &&
                methodInfo.GetGenericArguments().Length == 0 &&
                methodInfo.GetParameters().Length == 0)
            {
                // mocking IStockSnapshotProvider.GetLatestSnapshotAsync() (non-generic)
                //
                // This is the only method we support for mocking.
                return this.mockImpl;
            }
        }
        throw new NotImplementedException("This sample only supports mocking IStockSnapshotProvider.GetLatestSnapshotAsync()");
    }

    internal IStockSnapshotProvider Object
    {
        get { return this.mockImpl; }
    }
}

class MockSetup : IStockSnapshotProvider
{
    Action generalCallback = null;
    Func<Task<StockSnapshot>> valueCallback = null;
    Task<StockSnapshot> fixedValue = null;

    internal MockSetup Callback(Action action)
    {
        this.generalCallback = action;
        return this;
    }

    internal void Returns(Task<StockSnapshot> fixedValue)
    {
        this.fixedValue = fixedValue;
    }

    internal void Returns(Func<Task<StockSnapshot>> callback)
    {
        this.valueCallback = callback;
    }

    Task<StockSnapshot> IStockSnapshotProvider.GetLatestSnapshotAsync()
    {
        if (this.generalCallback != null)
        {
            this.generalCallback.Invoke();
        }

        if (this.valueCallback != null)
        {
            return valueCallback.Invoke();
        }

        if (this.fixedValue != null)
        {
            return fixedValue;
        }

        // we weren't set up with enough information to mock IStockSnapshotProvider
        throw new NotImplementedException("Not enough information to mock IStockSnapshotProvider");
    }
}

File: externals/Async-CTP/samples/VB/Testing/Unit/Testing/Test/Library/Mocks/Mock.vb
Imports System
Imports System.Linq.Expressions
Imports System.Threading.Tasks

Class Mock(Of TInterface)

    ReadOnly mockImpl As MockSetup = New MockSetup()

    Friend Sub New()
        If Not GetType(TInterface).[Equals](GetType(IStockSnapshotProvider)) Then
            Throw New NotImplementedException("This sample only supports mocking IStockSnapshotProvider")
        End If
    End Sub

    ' Our setup method only supports mocking IStockSnapshotProvider.GetLatestSnapshotAsync()
    Friend Function Setup([function] As Expression(Of Func(Of IStockSnapshotProvider, Task(Of StockSnapshot)))) As MockSetup
        If [function].Body.NodeType = ExpressionType.[Call] Then
            Dim innerCall = DirectCast([function].Body, MethodCallExpression)
            Dim methodInfo = innerCall.Method
            If GetType(IStockSnapshotProvider).[Equals](methodInfo.DeclaringType) AndAlso methodInfo.IsStatic = False AndAlso methodInfo.Name.[Equals]("GetLatestSnapshotAsync") AndAlso methodInfo.GetGenericArguments().Length = 0 AndAlso methodInfo.GetParameters().Length = 0 Then
                Return Me.mockImpl
            End If
        End If
        Throw New NotImplementedException("This sample only supports mocking IStockSnapshotProvider.GetLatestSnapshotAsync()")
    End Function

    Friend ReadOnly Property [Object] As IStockSnapshotProvider
        Get
            Return Me.mockImpl
        End Get
    End Property

End Class

Class MockSetup
    Implements IStockSnapshotProvider

    Dim generalCallback As Action = Nothing
    Dim valueCallback As Func(Of Task(Of StockSnapshot)) = Nothing
    Dim fixedValue As Task(Of StockSnapshot) = Nothing

    Friend Function Callback(action As Action) As MockSetup
        Me.generalCallback = action
        Return Me
    End Function

    Friend Sub Returns(fixedValue As Task(Of StockSnapshot))
        Me.fixedValue = fixedValue
    End Sub

    Friend Sub Returns(callback As Func(Of Task(Of StockSnapshot)))
        Me.valueCallback = callback
    End Sub

    Function GetLatestSnapshotAsync() As Task(Of StockSnapshot) Implements IStockSnapshotProvider.GetLatestSnapshotAsync
        If Me.generalCallback IsNot Nothing Then
            Me.generalCallback.Invoke()
        End If
        If Me.valueCallback IsNot Nothing Then
            Return valueCallback.Invoke()
        End If
        If Me.fixedValue IsNot Nothing Then
            Return fixedValue
        End If
        Throw New NotImplementedException("Not enough information to mock IStockSnapshotProvider")
    End Function

End Class